compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -6440,61 +6439,104 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
+          thisFormalCapture.toList ::: captureSyms.map(genParamDef(_, pos))
 
-        (thisFormalCapture :: formalCaptures,
-            body, thisActualCapture :: actualCaptures)
-      } else {
-        val body = genApplyStatic(target, allArgs)
+        // Gen the inlined target method body
+        val genMethodDef = {
+          genMethodWithCurrentLocalNameScope(consumeDelambdafyTarget(target),
+              initThisLocalVarIdent = thisFormalCapture.map(_.name))
+        }
+        val js.MethodDef(methodFlags, _, _, methodParams, _, methodBody) = genMethodDef
 
-        (formalCaptures, body, actualCaptures)
-      }
+        /* If the target method was not supposed to be static, but genMethodDef
+         * turns out to be static, it means it is a non-exposed method of a JS
+         * class. The `this` param was turned into a regular param, for which
+         * we need a `js.VarDef`.
+         */
+        val (maybeThisParamAsVarDef, remainingMethodParams) = {
+          if (methodFlags.namespace.isStatic && !isTargetStatic) {
+            val thisParamDef :: remainingMethodParams = methodParams: @unchecked
+            val thisParamAsVarDef = js.VarDef(thisParamDef.name, thisParamDef.originalName,
+                thisParamDef.ptpe, thisParamDef.mutable, thisFormalCapture.get.ref)
+            (thisParamAsVarDef :: Nil, remainingMethodParams)
+          } else {
+            (Nil, methodParams)
+          }
+        }
 
-      val (patchedFormalArgs, paramsLocals) =
-        patchFunParamsWithBoxes(target, formalArgs, useParamsBeforeLambdaLift = true)
+        // After that, the args found in the `Function` node had better match the remaining method params
+        assert(remainingMethodParams.size == allArgs0.size,
+            s"Arity mismatch: $remainingMethodParams <-> $allArgs0 at $pos")
 
-      val patchedBody =
-        js.Block(paramsLocals :+ ensureResultBoxed(body, target))
+        /* Declare each method param as a VarDef, initialized to the corresponding arg.
+         * In practice, all the args are `This` nodes of `VarRef` nodes, so the
+         * optimizer will alias those VarDefs away.
+         * We do it in two steps because we have different Symbols, hence different
+         * encoded LocalIdents.
+         */
+        val methodParamsAsVarDefs = for ((methodParam, arg) <- remainingMethodParams.zip(allArgs0)) yield {
+          js.VarDef(methodParam.name, methodParam.originalName, methodParam.ptpe,
+              methodParam.mutable, genExpr(arg))
+        }
 
-      val closure = js.Closure(
+        /* Adapt the params and result so that they are boxed from the outside.
+         * We need this because a `js.Closure` is always from `any`s to `any`.
+         */
+        val formalArgs = paramTrees.map(p => genParamDef(p.symbol))
+        val (patchedFormalArgs, paramsLocals) =
+          patchFunParamsWithBoxes(target, formalArgs, useParamsBeforeLambdaLift = true)
```

Comment:
This will definitely change again with #5003. That said per se the additional `VarDef`s will still be there, since types don't always align.

The changes in this PR are a status quo in terms of number of assignments that need to be processed by the optimizer: we traded `Binding` for `ParamDef`s in a method call for `Binding`s to local `VarDef`s.

We can look into improving this a bit further after #5003 if we think it's worth it.